networkx/algorithms/community/louvain.py

```diff
@@ -0,0 +1,110 @@
+"""Function for detecting communities base on Louvain Community Detection
+Algorithm"""
+
+from copy import deepcopy
+from random import sample
+
+import networkx as nx
+from networkx.algorithms.community import modularity
+
+__all__ = ["louvain_communities"]
+
+
+def louvain_communities(G, weight="weight", threshold=0):
+    """Find communities in G using the Louvain Community Detection
+    Algorithm.
+
+    Parameters
+    ----------
+    G : NetworkX graph
+    weight : string or None, optional (default="weight")
+        The name of an edge attribute that holds the numerical value
+        used as a weight. If None then each edge has weight 1.
+    threshold : float, optional (default=0)
+        Modularity gain threshold for each level. If the gain of modularity
+        between 2 levels of the algorithm is less than the given threshold
+        then the algorithm stops and returns the resulting communities.
+
+    Returns
+    -------
+    list
+        A list of lists of sets. Each sublist of sets represents one
+        level of the tree and each set represents the communities with
+        all the nodes that constitute them.
+
+    References
+    ----------
+    .. [1] Blondel, V.D. et al. Fast unfolding of communities in
+       large networks. J. Stat. Mech 10008, 1-12(2008)
+    """
+
+    partitions = []
+    partition = [{u} for u in G.nodes()]
+    mod = modularity(G, partition)
+    current_graph = G.copy()
+
+    while True:
+        partition, improvement = _one_level(current_graph, deepcopy(partition), weight)
+        if not improvement:
+            break
+        new_mod = modularity(G, partition)
+        if new_mod - mod <= threshold:
+            break
+        partitions.append(partition)
+        current_graph = _gen_graph(G, partition)
+    return partitions
+
+
+def _one_level(G, partition, weight="weight"):
+    """Calculate one level of the tree"""
+    node2com = {u: i for i, u in enumerate(G.nodes())}
+    total_weights = {i: G.degree(u, weight=weight) for i, u in enumerate(G.nodes())}
+    degrees = G.degree(weight=weight)
+    rand_nodes = sample(G.nodes, len(G.nodes))
+    nb_moves = 1
+    m = G.size(weight=weight)
```

Comment:
Minor points that probably won't change much... but might:

1) The `degrees` object is still a DegreeView of the graph, so it still has to compute the degree each time you look it up.  It might be faster to make that a dict:  `dict(G.degree(weight=weight)`.
2) Your `total_weights` can be initialized from `degrees` saving one pass of computing the degrees:
```python
degrees = dict(G.degree(weight=weight))
total_weights = {i: deg for i, deg in enumerate(degrees.values())}
```

3) The `rand_nodes` ordering can be constructed by [random.shuffle](https://docs.python.org/3/library/random.html#random.shuffle) which is built for this kind of operation, so should be faster. You have to form a list first because it changes the order in-place:  
```python
rand_nodes = random.shuffle(list(G.nodes))
```

The size of G (held in variable `m`) can be computed from `degrees` saving one more computation of the degree of each node.
```python
m = sum(degrees.values()) / 2  # for undirected graph "size" is half the sum of degrees
```

Also -- does m change as the communities change (and as we go from level to level?) It might be the same in which case we don't have to recompute it.

Reply:
I did make the proposed changes thanks.

Indeed the size `m` will always stay the same